target_chains/stylus/contracts/pyth-receiver/src/lib.rs

```diff
@@ -81,6 +164,96 @@ impl PythReceiver {
         // dummy implementation
     }
 
+    fn update_price_feeds_internal(&mut self, update_data: Vec<u8>) -> Result<(), PythReceiverError> {
+        let update_data_array: &[u8] = &update_data;
+        // Check the first 4 bytes of the update_data_array for the magic header
+        if update_data_array.len() < 4 {
+            return Err(PythReceiverError::InvalidUpdateData);
+        }
+
+        let mut header = [0u8; 4];
+        header.copy_from_slice(&update_data_array[0..4]);
+
+        if &header != PYTHNET_ACCUMULATOR_UPDATE_MAGIC {
+            return Err(PythReceiverError::InvalidAccumulatorMessage);
+        }
+
+        let update_data = AccumulatorUpdateData::try_from_slice(&update_data_array).unwrap();
+
+        match update_data.proof {
+            Proof::WormholeMerkle { vaa, updates } => {
+                let wormhole: IWormholeContract = IWormholeContract::new(self.wormhole.get());
+                let config = Call::new();
+                wormhole.parse_and_verify_vm(config, Vec::from(vaa.clone())).map_err(|_| PythReceiverError::InvalidWormholeMessage).unwrap();
+                
+                let vaa = Vaa::read(&mut Vec::from(vaa.clone()).as_slice()).unwrap();
+
+                let cur_emitter_address: &[u8; 32] = vaa.body.emitter_address.as_slice().try_into().expect("emitter address must be 32 bytes");
+
+                let cur_data_source = DataSource {
+                    chain_id: U16::from(vaa.body.emitter_chain),
+                    emitter_address: FixedBytes::from(cur_emitter_address),
+                };
+
+                if !self.is_valid_data_source.get(cur_data_source) {
+                    return Err(PythReceiverError::InvalidWormholeMessage);
+                }
+
+                let root_digest: MerkleRoot<Keccak160> = parse_wormhole_proof(vaa).unwrap();
```

Comment:
i recommend no unwrapping (or expecting) in contracts because it might get difficult to debug.